xc/xenops_server_xen.ml

```diff
@@ -3333,27 +3333,45 @@ module Actions = struct
     let vm = get_uuid ~xc domid |> Uuidm.to_string in
     Opt.iter
       (function { VmExtra.persistent } ->
-         if not persistent.VmExtra.pv_drivers_detected then begin
-           (* If the new value for this device is 4 then PV drivers are present *)
-           try
-             let value = xs.Xs.read path in
-             if value = "4" (* connected *) then begin
-               let updated =
-                 DB.update vm (
-                   Opt.map (function { VmExtra.persistent } ->
-                       let persistent = { persistent with VmExtra.pv_drivers_detected = true } in
-                       debug "VM = %s; found PV driver evidence on %s (value = %s)" vm path value;
-                       VmExtra.{persistent}
-                     )
-                 )
-               in
-               if updated then
-                 Updates.add (Dynamic.Vm vm) internal_updates
-             end
-           with Xs_protocol.Enoent _ ->
-             warn "Watch event on %s fired but couldn't read from it" path;
-             () (* the path must have disappeared immediately after the watch fired. Let's treat this as if we never saw it. *)
-         end
+        try
+          let value = xs.Xs.read path in
+          let pv_drivers_detected =
+            if value = "4" && not persistent.VmExtra.pv_drivers_detected then begin
+              (* State 4 ("connected") means that PV drivers are present for this device *)
+              debug "VM = %s; found PV driver evidence on %s (value = %s)" vm path value;
+              true
+            end else if value <> "4" && persistent.VmExtra.pv_drivers_detected then begin
+              (* This device is not connected, while earlier we detected PV drivers. *)
+              (* We conclude that drivers are still present if any other device is connected. *)
+              let devices = Device_common.list_frontends ~xs domid in
+              let found =
+                (* Return `true` as soon as a device in state 4 is found. *)
+                List.exists (fun device ->
+                  try
+                    xs.Xs.read (Device_common.backend_state_path_of_device ~xs device) = "4"
+                  with Xs_protocol.Enoent _ -> false
+                ) devices
+              in
+              if not found then (* No devices in state 4 *)
+                debug "VM = %s; lost PV driver evidence" vm;
+              found
```

Comment:
dodgy whitespace here

Reply:
No, the last line is not part of the if-statement, which is only for logging.